Packet++/src/S7CommLayer.cpp

```diff
@@ -0,0 +1,117 @@
+#include "EndianPortable.h"
+
+#include "S7CommLayer.h"
+#include "TcpLayer.h"
+#include <iostream>
+
+#include <cstring>
+#include <sstream>
+
+namespace pcpp
+{
+
+	S7CommLayer::S7CommLayer(uint8_t msgType, uint16_t pduRef, uint16_t paramLength, uint16_t dataLength,
+							 uint8_t errorClass, uint8_t errorCode)
+	{
+		size_t headerLen;
+		if (msgType == 0x03)
+		{
+			headerLen = sizeof(s7comm_ack_data_hdr);
+			m_DataLen = headerLen;
+			m_Data = new uint8_t[headerLen];
+			memset(m_Data, 0, headerLen);
+			s7comm_ack_data_hdr *ack_d = (s7comm_ack_data_hdr *)m_Data;
+			ack_d->protocolId = 0x32;
+			ack_d->msgType = msgType;
+			ack_d->reserved = 0x0000;
+			ack_d->pduRef = htobe16(pduRef);
+			ack_d->paramLength = htobe16(paramLength);
+			ack_d->dataLength = htobe16(dataLength);
+			ack_d->errorClass = errorClass;
+			ack_d->errorCode = errorCode;
+		}
+		else
+		{
+			headerLen = sizeof(s7commhdr);
+			m_DataLen = headerLen;
+			m_Data = new uint8_t[headerLen];
+			memset(m_Data, 0, headerLen);
+			s7commhdr *s7commHdr = (s7commhdr *)m_Data;
+			s7commHdr->protocolId = 0x32;
+			s7commHdr->msgType = msgType;
+			s7commHdr->reserved = 0x0000;
+			s7commHdr->pduRef = htobe16(pduRef);
+			s7commHdr->paramLength = htobe16(paramLength);
+			s7commHdr->dataLength = htobe16(dataLength);
+		}
+
+		m_Protocol = S7COMM;
+	}
+
+	std::string S7CommLayer::toString() const
+	{
+		std::ostringstream str;
+
+		str << "S7Comm Layer, Job Request";
+
+		return str.str();
+	}
+
+	bool S7CommLayer::isDataValid(const uint8_t *data, size_t dataSize)
+	{
+		if (!data || dataSize < sizeof(s7commhdr))
+			return false;
+
+		return data[0] == 0x32;
+	}
+
+	uint8_t S7CommLayer::getProtocolId() const { return getS7commHeader()->protocolId; }
+
+	uint8_t S7CommLayer::getMsgType() const { return getS7commHeader()->msgType; }
+
+	uint16_t S7CommLayer::getParamLength() const { return be16toh(getS7commHeader()->paramLength); }
+
+	uint16_t S7CommLayer::getPduRef() const { return be16toh(getS7commHeader()->pduRef); }
+
+	uint16_t S7CommLayer::getDataLength() const { return be16toh(getS7commHeader()->dataLength); }
+
+	void S7CommLayer::setMsgType(uint8_t msgType) const { getS7commHeader()->msgType = msgType; }
+
+	uint8_t S7CommLayer::getErrorCode() const { return getS7commAckDataHeader()->errorCode; }
+
+	uint8_t S7CommLayer::getErrorClass() const { return getS7commAckDataHeader()->errorClass; }
+
+	void S7CommLayer::setParamLength(uint16_t paramLength) const
+	{
+		getS7commHeader()->paramLength = htobe16(paramLength);
```

Comment:
If the user changes `paramLength`, we should also change the length of the parameter data, no?
Otherwise the packet will become invalid

The problem is what should happen if there is existing parameter data and the user changes its length (increases or decreases the length)?

I see 2 options:
1. Let the user set the parameter data, something like `s7CommLayer->setParameter(data, dataLen)`
2. Don't let the user set the parameter data, and remove this `setParamLength()` method

The second option is obviously easier than the first one, but I think both options are ok. Which one do you prefer?